src/org/rascalmpl/dap/RascalDebugAdapter.java

```diff
@@ -506,48 +521,221 @@ public CompletableFuture<Void> pause(PauseArguments args) {
+        detail.append(func.toString());
+        if (func instanceof NamedFunction) {
+            NamedFunction nf = (NamedFunction) func;
+            if (nf.hasTag("synopsis")) {
+                IValue v = nf.getTag("synopsis");
+                if (v instanceof IString) {
+                    detail.append(((IString) v).getValue());
+                    detail.append("\n");
+                }
+            }
+        }
+        return detail.toString();
+    }
+
+    private void addVariablesCompletionsFromFrame(IRascalFrame frame, String text, List<CompletionItem> completions) {
+        Set<String> frameVariables = frame.getFrameVariables();
+        for(String var : frameVariables){
+            if(var.startsWith(text)){
+                CompletionItem completion = new CompletionItem();
+                IRascalResult varInFrame = frame.getFrameVariable(var);
+                completion.setLabel(var+" : " + varInFrame.getDynamicType().toString());
+                completion.setSortText(var);
+                if (text.length() < var.length()) { //remove the prefix
+                    completion.setText(var.substring(text.length()));
+                }
+                else { // exact match
+                    completion.setText("");
+                }
+                completion.setType(CompletionItemType.VARIABLE);
+                completions.add(completion);
+            }
+        }
+    }
+
+    private void addFunctionsCompletionsFromEnvironment(Environment env, String text, List<CompletionItem> completions) {
+        for(Pair<String,LinkedHashSet<AbstractFunction>> functions : env.getFunctions()){
+            String funcName = functions.getFirst();
+            if(funcName.startsWith(text)){
+                for(AbstractFunction func : functions.getSecond()){
+                    CompletionItem completion = new CompletionItem();
+                    completion.setLabel(func.getHeader());
+                    completion.setSortText(funcName);
+                    if (text.length() < funcName.length()) { //remove the prefix
+                        completion.setText(funcName.substring(text.length()));
+                    }
+                    else { // exact match
+                        completion.setText("");
+                    }
+                    completion.setType(CompletionItemType.FUNCTION);
+                    completion.setDetail(getFunctionDetail(func));
+                    completions.add(completion);
+                }
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<CompletionsResponse> completions(CompletionsArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            CompletionsResponse response = new CompletionsResponse();
+            List<CompletionItem> completions = new ArrayList<>();
+            if(suspendedState.isSuspended()) {
+                Integer frameId = args.getFrameId();
+                assert frameId != null;
```

Comment:
is this for us developers? in that case add a message to the assert: `assert frameId != null : "at this point there should never be a missing frameid, as we always pass in a correct one";` or something like this.

otherwise it should be an if + exception